### PR TITLE
Colorize command parameters and privilege names

### PR DESCRIPTION
**Pull request by Panquesito7**
Originally by @Wuzzy2.
See #7019 for screenshots and information.

(I did not split changes into separate commits, forgot to do that, apologizes.)

## To do
This PR is ready for review.

### CODE REVIEW

**Comment by paramat** [comment]
1 commit is fine.
Are the changes identical to #7019 apart from any necessary rebasing?
Note that SmallJoker approved the original PR, and i approve the concept.

**Comment by Panquesito7** [comment]
Changes are the same, nothing changed (besides rebase).

**Comment by Panquesito7** [comment]
> Note that SmallJoker approved the original PR

Shouldn't the `One approval` label be added?

> and i approve the concept.

Does this count as an second approval?

**Comment by paramat** [comment]
Because it is a new PR based on altered code it should ideally be rechecked by @SmallJoker before their approval is given.

**Comment by Panquesito7** [comment]
Ping?

**Comment by SmallJoker** [comment]
Concept is good (still), but there are a few stylistic errors. I also still think that this formatting could be made easier by implementing a `string.format`-like function: (untested, concept)
```Lua
local example = core.colorize_chatcommand(
	"@cmd [@arg | @arg]", "/teleport", "<nickname>", "(X, Y, Z)")

function core.colorize_chatcommand(str, ...)
	local args = { ... }
	local str_pos = 1
	for i = 1, #args do
		local pos_cmd = str:find("@cmd", str_pos, true)
		local pos_arg = str:find("@arg", str_pos, true)
		if pos_cmd and (not pos_arg or pos_cmd < pos_arg) then
			-- replace "@cmd" with colorized args[i]
		elseif pos_arg then
			-- replace "@arg" with colorized args[i]
		else
			break -- No further occurences
		end
	end
	return str
end
```

**Comment by Panquesito7** [comment]
> but there are a few stylistic errors

Which ones? Are they reviewed in the old PR?

File: builtin/common/chatcommands.lua [hunk 1]
```diff
@@ -31,26 +31,34 @@ end
...
-		local msg = core.colorize("#00ffff", cmd_marker .. cmd)
-		if def.params and def.params ~= "" then
-			msg = msg .. " " .. def.params
+	local function format_help_line(cmd, def, is_privilege)
+			local msg
+			if is_privilege then
+				msg = core.colorize(core.COLOR_PRIV, cmd)
+			else
+				cmd = cmd_marker .. cmd
+				msg = core.colorize_chatcommand(cmd, def.params)
```
Comment (SmallJoker) [inline builtin/common/chatcommands.lua]: Code block above: -1 indent.

File: builtin/common/chatcommands.lua [hunk 2]
```diff
@@ -65,9 +73,17 @@ local function do_help_cmd(name, param)
...
-				.. gettext_replace("Use '$1help <cmd>' to get more information,"
-				.. " or '$1help all' to list everything.", cmd_marker)
+		-- Intentionally not translated because the cmd names are static
+		local help_cmd = core.colorize_chatcommand(
+				string.format("%shelp", cmd_marker), "<cmd>")
+		local help_all = core.colorize_chatcommand(
+				string.format("%shelp", cmd_marker), "all")
+		-- List available commands
+		return true, gettext("Available commands: ")
+				.. core.colorize(core.COLOR_COMMAND, table.concat(cmds, " "))
```
Comment (SmallJoker) [inline builtin/common/chatcommands.lua]: One indent is enough, move binary operators to the end of the previous line
.

File: builtin/common/misc_helpers.lua [hunk 3]
```diff
@@ -642,6 +643,13 @@ function core.colorize(color, message)
 	return table.concat(lines, "\n") .. core.get_color_escape_sequence("#ffffff")
 end
 
+function core.colorize_chatcommand(cmd, param)
+	if param ~= nil and param ~= "" then
+		return core.colorize(core.COLOR_COMMAND, cmd) .. " " .. core.colorize(core.COLOR_PARAM, param)
```
Comment (SmallJoker) [inline builtin/common/misc_helpers.lua]: Line too long

File: builtin/game/chat.lua [hunk 4]
```diff
@@ -126,8 +143,7 @@ core.register_chatcommand("privs", {
 			return false, "Player " .. name .. " does not exist."
 		end
 		return true, "Privileges of " .. name .. ": "
-			.. core.privs_to_string(
-				core.get_player_privs(name), ' ')
+		.. format_player_privs(name)
```
Comment (SmallJoker) [inline builtin/game/chat.lua]: +1 indent.

File: builtin/game/chat.lua [hunk 5]
```diff
@@ -194,11 +210,10 @@ local function handle_grant_command(caller, grantname, grantprivstr)
 	if grantname ~= caller then
 		core.chat_send_player(grantname, caller
 				.. " granted you privileges: "
-				.. core.privs_to_string(grantprivs, ' '))
+				.. format_privs(grantprivs))
 	end
 	return true, "Privileges of " .. grantname .. ": "
-		.. core.privs_to_string(
-			core.get_player_privs(grantname), ' ')
+	.. format_player_privs(grantname)
```
Comment (SmallJoker) [inline builtin/game/chat.lua]: +1 indent. Same in multiple places below. Please check your changes.

File: builtin/game/chat.lua [hunk 6]
```diff
@@ -889,7 +932,7 @@ core.register_chatcommand("days", {
 
 core.register_chatcommand("shutdown", {
 	params = "[<delay_in_seconds> | -1] [reconnect] [<message>]",
-	description = "Shutdown server (-1 cancels a delayed shutdown)",
+	description = "Shutdown server ("..core.colorize(core.COLOR_PARAM, "-1") .." cancels a delayed shutdown)",
```
Comment (SmallJoker) [inline builtin/game/chat.lua]: Line too long

File: doc/lua_api.txt [hunk 7]
```diff
@@ -4000,7 +4009,7 @@ Call these functions only at load time!
 * `minetest.override_item(name, redefinition)`
     * Overrides fields of an item registered with register_node/tool/craftitem.
     * Note: Item must already be defined, (opt)depend on the mod defining it.
-    * Example: `minetest.override_item("default:mese",
+    * Example: `minetest.override_item("default:mese",
```
Comment (SmallJoker) [inline doc/lua_api.txt]: Erroneous change.

**Comment by Panquesito7** [comment]
Thanks for your review, @SmallJoker. 🙂
I've created a commit to fix them; hope it's alright.

**Comment by Panquesito7** [comment]
@SmallJoker.

**Comment by nerzhul** [comment]
please rebase

**Comment by Panquesito7** [comment]
Rebased.

File: builtin/client/chatcommands.lua [hunk 8]
```diff
@@ -8,7 +8,7 @@ core.register_on_sending_chat_message(function(message)
 
 	local first_char = message:sub(1,1)
 	if first_char == "/" or first_char == "." then
-		core.display_chat_message(core.gettext("issued command: ") .. message)
+		core.display_chat_message(core.gettext("Issued command: ") .. message)
```
Comment (SmallJoker) [inline builtin/client/chatcommands.lua:11]: unrelated change.

File: builtin/common/chatcommands.lua [hunk 9]
```diff
@@ -65,9 +73,16 @@ local function do_help_cmd(name, param)
...
-		return true, gettext("Available commands: ") .. table.concat(cmds, " ") .. "\n"
-				.. gettext_replace("Use '$1help <cmd>' to get more information,"
-				.. " or '$1help all' to list everything.", cmd_marker)
+		-- Intentionally not translated because the cmd names are static
+		local help_cmd = core.colorize_chatcommand(
+				string.format("%shelp", cmd_marker), "<cmd>")
+		local help_all = core.colorize_chatcommand(
+				string.format("%shelp", cmd_marker), "all")
+		-- List available commands
+		return true, gettext("Available commands: ") .. core.colorize(core.COLOR_COMMAND, table.concat(cmds, " "))
```
Comment (SmallJoker) [inline builtin/common/chatcommands.lua:82]: Line too long.

File: builtin/common/misc_helpers.lua [hunk 10]
```diff
@@ -587,6 +587,14 @@ function core.colorize(color, message)
 	return table.concat(lines, "\n") .. core.get_color_escape_sequence("#ffffff")
 end
 
+function core.colorize_chatcommand(cmd, param)
+	if param ~= nil and param ~= "" then
+		return core.colorize(core.COLOR_COMMAND, cmd) .. " " 
+		.. core.colorize(core.COLOR_PARAM, param)
```
Comment (SmallJoker) [inline builtin/common/misc_helpers.lua:593]: Missing indent.

File: builtin/common/misc_helpers.lua [hunk 11]
```diff
@@ -587,6 +587,14 @@ function core.colorize(color, message)
...
+	else
```
Comment (SmallJoker) [inline builtin/common/misc_helpers.lua:594]: Early return = superfluous `else` case
```suggestion
	end
```
And drop the `end` below.

File: builtin/game/chat.lua [hunk 12]
```diff
@@ -1,5 +1,14 @@
 -- Minetest: builtin/game/chat.lua
 
+-- Pretty-print list of privileges
+local function format_privs(privs)
+	return core.colorize(core.COLOR_PRIV, core.privs_to_string(privs, ", "))
+end
+
+-- Pretty-print privileges of give player
+local function format_player_privs(name)
+	return format_privs(core.get_player_privs(name))
```
Comment (SmallJoker) [inline builtin/game/chat.lua:10]: Missing end.

File: builtin/game/chat.lua [hunk 13]
```diff
@@ -71,9 +80,14 @@ core.register_on_chat_message(function(name, message)
 			core.chat_send_player(name, result)
 		end
 	else
+		local privs = {}
+		for i=1, #missing_privs do
+			privs[missing_privs[i]] = true
+		end
+		local priv_string = format_privs(privs)
```
Comment (SmallJoker) [inline builtin/game/chat.lua:87]: Redundant temporary variable. Consider using `format_privs(privs)` in-line below.

File: builtin/game/chat.lua [hunk 14]
```diff
@@ -448,7 +467,8 @@ core.register_chatcommand("teleport", {
 		end
 
 		if not core.check_player_privs(name, {bring=true}) then
-			return false, "You don't have permission to teleport other players (missing bring privilege)"
+			return false, "You don't have permission to teleport other players (missing "
+			.. core.colorize(core.COLOR_PRIV, "bring") .. " privilege)"
```
Comment (SmallJoker) [inline builtin/game/chat.lua:471]: Missing indent.

File: builtin/game/chat.lua [hunk 15]
```diff
@@ -904,7 +946,8 @@ core.register_chatcommand("days", {
 
 core.register_chatcommand("shutdown", {
 	params = "[<delay_in_seconds> | -1] [reconnect] [<message>]",
-	description = "Shutdown server (-1 cancels a delayed shutdown)",
+	description = "Shutdown server ("..core.colorize(core.COLOR_PARAM, "-1") ..
+	" cancels a delayed shutdown)",
```
Comment (SmallJoker) [inline builtin/game/chat.lua:950]: Missing indent.

File: builtin/game/chat.lua [hunk 16]
```diff
@@ -904,7 +946,8 @@ core.register_chatcommand("days", {
 
 core.register_chatcommand("shutdown", {
 	params = "[<delay_in_seconds> | -1] [reconnect] [<message>]",
-	description = "Shutdown server (-1 cancels a delayed shutdown)",
+	description = "Shutdown server ("..core.colorize(core.COLOR_PARAM, "-1") ..
```
Comment (SmallJoker) [inline builtin/game/chat.lua:949]: Spaces around binary operators: `a .. b`

Same for the following changes in this function.

File: doc/lua_api.txt [hunk 17]
```diff
@@ -2903,6 +2909,9 @@ The following functions provide escape sequences:
       `minetest.get_color_escape_sequence(color) ..
       message ..
       minetest.get_color_escape_sequence("#ffffff")`
+   `minetest.colorize_chatcommand(cmd, param)`
```
Comment (SmallJoker) [inline doc/lua_api.txt:2912]:
```suggestion
* `minetest.colorize_chatcommand(cmd, param)`
```

**Comment by Panquesito7** [comment]
Rebased.

**Comment by Panquesito7** [comment]
@SmallJoker ping; is the PR fine now?

**Comment by Panquesito7** [comment]
@SmallJoker No pressure, but, could you please re-review?

**Comment by Panquesito7** [comment]
@SmallJoker ping?

**Comment by Panquesito7** [comment]
Ping? It's been almost a year since I authored this.
When will it be merged? Is there anything else to be addressed?

**Comment by paramat** [comment]
Sorry for the delay.
All comments seem resolved so it seems only reviewer time is the issue.

**Comment by Panquesito7** [comment]
It's now been a year since I authored this PR. Who will review this? @SmallJoker. Anyone else?

**Comment by Panquesito7** [comment]
@paramat @SmallJoker ping? 😕

File: builtin/game/chat.lua [hunk 18]
```diff
@@ -628,8 +661,11 @@ core.register_chatcommand("deleteblocks", {
 
 core.register_chatcommand("fixlight", {
 	params = "(here [<radius>]) | (<pos1> <pos2>)",
-	description = "Resets lighting in the area between pos1 and pos2 "
-		.. "(<pos1> and <pos2> must be in parentheses)",
+	description = "Resets lighting in the area from "
+		.. core.colorize(core.COLOR_PARAM, "<pos1>").." to "
```
Comment (MoNTE48) [inline builtin/game/chat.lua:665]: pls add space before and after `..` :)

Comment (Panquesito7) [reply builtin/game/chat.lua]: Done.

**Comment by Zughy** [comment]
@Panquesito7 the issue is now "Supported by core dev". I understand the frustration, but I must ask: are you going to rebase this or should I close it (as the rebase request is from 27 February 2021)?

**Comment by Panquesito7** [comment]
> @Panquesito7 the PR is now "Supported by core dev". I understand the frustration, but I must ask: are you going to rebase this or should I close it (as the rebase request is from 27 February 2021)?

Hey there! I'll look into it on these days and see if I can rebase it. If not, I'll make another PR and will notify you about it here. Thanks for the reminder. :)

**Comment by Zughy** [comment]
@Panquesito7 please consider opening another PR, closing here in the meanwhile (30 days have passed + ping on Matrix that went unnoticed)
